scalalib/src/mill/scalalib/JavaModule.scala

```diff
@@ -847,10 +847,61 @@ trait JavaModule
    */
   def generatedSources: T[Seq[PathRef]] = Task { Seq.empty[PathRef] }
 
+  /**
+   * A set of tasks producing generated code. The difference between this and
+   * [[generatedSources]] lies in the fact that IDE-related tasks
+   * (BSP/Bloop/GenIdea) will not run the tasks, but will use them to statically
+   * determine the location of the source code they will produce by looking up their
+   * [[Task#generatedSourceRoots]], knowing that these roots will be present under
+   * the dest directories once the task is run.
+   */
+  def sourceGenerators: Seq[NamedTask[_]] = Seq.empty
+
+  private def predictGeneratedSourceRoots(task: NamedTask[_])(workspace: os.Path): Seq[os.Path] = {
+    val dest = mill
+      .eval
+      .EvaluatorPaths
+      .resolveDestPaths(workspace / "out", task)
+      .dest
+    val explicitSourceRoots = task.generatedSourceRoots
+    if (explicitSourceRoots.isEmpty) Seq(dest)
+    else explicitSourceRoots.map(subPath => dest / subPath)
+  }
+
+  /**
+   * Computes the list of source roots that will be produced by [[sourceGenerators]] without
+   * actually running the generators in question.
+   */
+  def predictedGeneratedSourceRoots: T[Seq[PathRef]] = T {
+    val ws = T.workspace
+    sourceGenerators.flatMap(t => predictGeneratedSourceRoots(t)(ws)).map(PathRef(_))
+  }
+
+  /**
+   * Runs the lazy source generators, returning references to the expanded generated source roots
+   * they are supposed to have written code in.
+   */
+  final def generatedSourceRoots: T[Seq[PathRef]] = T {
```

Comment:
We can call this `deferredGeneratedSources`, to be in line with `sources`, `generatedSources`, `allSources`, etc.